client/components/site-selector/index.jsx

```diff
@@ -165,122 +300,97 @@ const SiteSelector = React.createClass( {
+					onSelect={ this.onAllSitesSelect }
+					onMouseEnter={ this.onAllSitesHover }
+					isHighlighted={ isHighlighted }
+					isSelected={ this.isSelected( ALL_SITES ) }
+					ref={ isHighlighted ? 'highlightedSite' : null }
 				/>
 			);
 		}
 	},
 
-	renderRecentSites() {
-		if ( this.state.search || ! this.shouldShowGroups() ) {
-			return;
-		}
+	renderSite( site ) {
+		this.visibleSites.push( site );
+
+		const isHighlighted = this.isHighlighted( site );
+		return (
+			<Site
+				site={ site }
+				key={ 'site-' + site.ID }
+				indicator={ this.props.indicator }
+				onSelect={ this.onSiteSelect }
+				onMouseEnter={ this.onSiteHover }
+				isHighlighted={ isHighlighted }
+				isSelected={ this.isSelected( site ) }
+				ref={ isHighlighted ? 'highlightedSite' : null }
+			/>
+		);
+	},
 
+	renderRecentSites() {
 		const sitesById = keyBy( this.props.sites.get(), 'ID' );
+		const sites = this.props.recentSites.map( siteId => sitesById[ siteId ] );
 
-		return (
-			<div className="site-selector__recent">
-				{ map( this.props.recentSites, ( siteId ) => {
-					const site = sitesById[ siteId ];
-					if ( ! site ) {
-						return;
-					}
+		if ( ! sites || this.state.search || ! this.shouldShowGroups() || this.props.visibleSiteCount <= 11 ) {
+			return null;
+		}
 
-					let siteHref;
-					if ( this.props.siteBasePath ) {
-						siteHref = this.getSiteBasePath( site ) + '/' + site.slug;
-					}
+		const recentSites = sites.map( this.renderSite, this );
 
-					return (
-						<Site
-							site={ site }
-							href={ siteHref }
-							key={ 'site-' + site.ID }
-							indicator={ this.props.indicator }
-							onSelect={ this.onSiteSelect.bind( this, site.slug ) }
-							isSelected={ this.isSelected( site ) }
-						/>
-					);
-				} ) }
-			</div>
-		);
+		if ( ! recentSites ) {
+			return null;
+		}
+
+		return <div className="site-selector__recent">{ recentSites }</div>;
 	},
 
 	render() {
 		const hiddenSitesCount = this.props.siteCount - this.props.visibleSiteCount;
 		const selectorClass = classNames( 'site-selector', 'sites-list', {
-			'is-large': this.props.siteCount > 6 || hiddenSitesCount > 0,
-			'is-single': this.props.visibleSiteCount === 1
+			'is-large': this.props.siteCount > 6 || hiddenSitesCount > 0 || this.state.showSearch,
+			'is-single': this.props.visibleSiteCount === 1,
+			'is-hover-enabled': ! this.state.isKeyboardEngaged,
 		} );
 
+		this.visibleSites = [];
+
 		return (
-			<div className={ selectorClass }>
+			<div className={ selectorClass } onMouseMove={ this.onMouseMove } onMouseLeave={ this.onMouseLeave }>
 				<Search
 					ref="siteSearch"
 					onSearch={ this.onSearch }
 					autoFocus={ this.props.autoFocus }
 					disabled={ ! this.props.sites.initialized }
 					onSearchClose={ this.props.onClose }
+					onKeyDown={ this.onKeyDown }
 				/>
 				<div className="site-selector__sites" ref="selector">
 					{ this.renderAllSites() }
+					{ this.renderRecentSites() }
```

Comment:
I'm sure we can move out `renderNewSiteButton` and `hidden-sites-message`, but `renderAllSites` `renderRecentSites` & `renderSites` should stay here, as the are sharing a lot of state and they also work with off-state `this.visibleSites` that we use for keyboard accessibility. It would be difficult to make this all sync across more components.